pybossa/api/__init__.py

```diff
@@ -461,3 +464,52 @@ def task_gold(project_id=None):
     task_repo.update(task)
 
     return Response(json.dumps({'success': True}), 200, mimetype="application/json")
+
+
+@jsonpify
+@login_required
+@csrf.exempt
+@blueprint.route('/task/<task_id>/services/<service_name>/<service_version>', methods=['POST'])
+@ratelimit(limit=ratelimits.get('LIMIT'), per=ratelimits.get('PER'))
+def get_service_request(task_id, service_name, service_version):
+    """Proxy service call"""
+    proxy_service_config = current_app.config.get('PROXY_SERVICE_CONFIG', None)
+    task = task_repo.get_task(task_id)
+    project = project_repo.get(task.project_id)
+
+    if not task or not proxy_service_config or not service_name or not service_version:
+        return abort(400)
+
+    timeout = project.info.get('timeout', ContributionsGuard.STAMP_TTL)
+    task_locked_by_user = has_lock(task.id, current_user.id, timeout)
+    payload = request.json
+
+    if task_locked_by_user:
+        service = _get_valid_service(task_id, service_name, service_version, payload, proxy_service_config) if isinstance(payload, dict) else None
```

Comment:
this would work. a minor suggestion is to consolidate all param checks under get_valid_service including payload, so
> service = _get_valid_service(task_id, service_name, service_version, payload, proxy_service_config)
> if 'headers' in service:
> .          url = ...

Reply:
Why do we need to check  if headers exist? that is part of our configurations on settings_local, we should add them when we configure a new service. In respect to Uri is the same for all the services. If you check in the settings local uri is not part of the service it self.

Reply:
as the func name is _get_valid_service, potentially all validation can happen inside this func including checking for headers before accessing it, so that exceptions that can occur due to bad input can handled gracefully as in certain cases, such endpoint can be called quite often such as each key press or so...

Reply:
But currently all validations are in that function everything else is just calling the service. Which lines exactly do you think we should  move  in to that function?

Reply:
as header are from config, checking header can be avoided. I think line 483 - 485 (https://github.com/bloomberg/pybossa/pull/309/files#diff-e23f749f9cbf6993dd73b23a99f4707dR483) can be part of _get_valid_service. However, you can take a call at your discretion looking at code maintainability; either would work. Thanks.

Reply:
I will actually use get('headers') to ensure if the headers are not define on our config this doesn't break. The other lines will keep them where they are.